orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py

```diff
@@ -622,6 +647,43 @@ def __setstate__(self, state):
 
         _utils.reinitialize_graph_execution_manager(self)
 
+    def _check_embedding_sparsity(self):
+        if not self._runtime_options.enable_embedding_sparse_optimizer or self._device.type != "cuda":
+            return
+
+        def embedding_hook(module, args, output):
+            ebd_input = args[0]
+            if ebd_input is None or not isinstance(ebd_input, torch.Tensor):
+                self._logger.warning("Embedding input is not a tensor.")
+                return None
+
+            valid_token = torch.count_nonzero(ebd_input - module.padding_idx)
+            total_token = ebd_input.numel()
+            embed_density = float(valid_token) / float(total_token) * 100
+            if embed_density < 90:
+                self._logger.info("Embedding sparsity-based optimization is ON for density: %.0f%%", embed_density)
+                if module not in self._embedding_module_to_padding_density_map:
+                    self._logger.warning("Found Embedding module not in the map. %s", module)
+                    return None
+                if (
+                    module in self._embedding_module_to_padding_density_map
+                    and self._embedding_module_to_padding_density_map[module][1] != -1
+                ):
+                    self._logger.warning(
+                        "Found duplicate Embedding module. %s", self._embedding_module_to_padding_density_map[module][0]
+                    )
+                self._embedding_module_to_padding_density_map[module][1] = embed_density
+                return _FlagPaddingElimination.apply(output)
+            else:
+                self._logger.info("Embedding sparsity-based optimization is OFF for density: %.0f%%", embed_density)
+                return None
+
+        for name, sub_module in self._flattened_module.named_modules():
+            if isinstance(sub_module, torch.nn.modules.sparse.Embedding):
+                if sub_module.padding_idx is not None and sub_module.padding_idx >= 0:
+                    self._embedding_module_to_padding_density_map[sub_module] = [name, -1]
+                    sub_module.register_forward_hook(embedding_hook)
```

Comment:
better to remove the hook after the model export or exception catching branch if it is failed.

Reply:
this hook must be added before model export.